### ISSUE TITLE
Questions about Audi A8 Seat Control

### ISSUE THREAD

**Issue by artemdevo** [k7000i/peugeot508CAN#1]
Hello, 

Sorry for opening an issue on this repo, but I wanted to try and contact you somehow. I found your posts on this AudiWorld thread: https://www.audiworld.com/forums/a8-s8-d4-platform-discussion-190/audi-a8-seat-chair-2919108/page5/

I  have a multicontour seat from a 2014 A8, and I am trying to do basically the same thing with an mcp2515 arduino module, however the seat does not reply to any CAN messages I send, it only pulses the bus every 50 microseconds.

I would love to hear more about how you implemented this. 

Thank you so much.

**Comment by k7000i** [comment]
Hello, nice to hear about that project :D

In my case i had to send the ignition can bus message in order to wake up the modules.
The code is the following:

`      SPI_init(SPI_4MHz);
        reset();
        setCanSpeed(CAN_500kbps);
        setCanMode(CANMODE_NORMAL);

        
        _delay_us(500);
        canFrameTx.can_id = 0x3C0;
        canFrameTx.can_extended = 0;
        canFrameTx.can_dlc = 4;
        canFrameTx.can_data[0] = 0;
        canFrameTx.can_data[1] = 0;
        canFrameTx.can_data[2] = 0xff;
        canFrameTx.can_data[3] = 0xff;
        sendMessage(&canFrameTx);

        _delay_us(1000);
        canFrameTx.can_id = 0x664;
        canFrameTx.can_extended = 0;
        canFrameTx.can_dlc = 6;
        canFrameTx.can_data[0] = 0;
        canFrameTx.can_data[1] = seat_hvac;
        canFrameTx.can_data[2] = seat_hvac;
        canFrameTx.can_data[3] = 0;
        canFrameTx.can_data[4] = 0;
        canFrameTx.can_data[5] = 0;
        sendMessage(&canFrameTx);`

The seat hvac byte first nibble is the cooling fan setting 0x1,0x4,0xD the second nibble is the heating setting 0x0E,0x0C,0x06

The pneumatics are controlled directly with the arduio from the onboard LM9733 chip

**Comment by artemdevo** [comment]
Thank you so much! I see what messages you are trying to send but the only library I have working with my mcp2515 module doesn't work with 500 kbps. Can you please post all of the code that you loaded onto the arduino? I would be incredibly thankful. I tried to look at the code you made for the Peugeot but it won't work easily with the arduino IDE.

![imagecan](https://github.com/user-attachments/assets/54f14197-80d1-4db2-bf15-da18ac7e79f8)
The picture is of my module transmitting the wake up, message, at 100 kbps. The seat module does not acknowledge it

**Comment by k7000i** [comment]
You need to tweak the cfg bits on mcp2515 depending on your arduino crystal frequency. Keep the wires relatively short. You can use the following site to calculate the bits.
https://kvaser.com/support/calculators/bit-timing-calculator/
My settings are:
```
cfg1 = 0x00;
cfg2 = 0x90;
cfg3 = 0x02;
```

**Comment by artemdevo** [comment]
Hi again, 

So I got the HVAC stuff working. In the AudiWorld posts, you mention that you did get the massage to work, in addition to the side cushions, although you ultimately implemented massage with direct arduino control. Here is the post:

![k7000i](https://github.com/user-attachments/assets/172061ff-d7e4-494a-9c51-fc5d8978eaa4)

Can you explain these IDs or show the code more please? I would like to try and use CAN for as much as possible.

Also you say you "brute force" the CAN IDs. I didn't think that was possible, so could you please explain how you found these? 

Thank you so much.

**Comment by k7000i** [comment]
Yeah, i tought i can brute force the id-s but no joy, the component protection kicks in and then game over. So the only option is to sniff the can on a real A8 or A6. That would be a good starting point, but i think maybe some difference in different hardware revisions. I also tried simulate a car with a canbus gateway modul an odb port and the seat with original diagnostic equipment, but no joy. I also tried digging deep in the infotainment system and reverse engineered a bit, but also no luck. So i wish you a good luck :)

**Comment by k7000i** [comment]
Yes those id-s are worked for 10 minutes, then i pressed the massage button on the seat and never worked again. I assume the component protection is activated.

**Comment by k7000i** [comment]
If you worked it out the canbus msg-s let me know ;). I have the passanger side seat spared for this exact purpose, but i havent had access to a real car yet.

**Comment by artemdevo** [comment]
Ok thank you. So where did you happen to find the HVAC ID, 0x664 from then?

Additionally, do you think you could describe how you implemented the pneumatics in more detail? It would just make things a bit easier when I go to do it.
![image](https://github.com/user-attachments/assets/722f0a0a-f54a-4e6d-999e-31219c38f0ff)
 You mention a vent valve solenoid here. Can you tell me which one that is? It looks like all of the solenoids go to a bladder, that I can see.

**Comment by artemdevo** [comment]
Hello, its me again! Aside from the hvac can frame, I ended up controlling everything manually. I never got any other CAN ID's for anything. However, I am still very curious where you got the HVAC frame ID, 0x664 from.

I would really appreciate your reply, thank you!
